tests/unit/mesh/data/test_gridfromoptions.cxx

```diff
@@ -254,43 +292,202 @@ TEST_F(GridFromOptionsTest, GetVectorBoutRealZ) {
   EXPECT_EQ(result, expected);
 }
 
 TEST_F(GridFromOptionsTest, GetVectorBoutRealZNone) {
   std::vector<BoutReal> result{};
   std::vector<BoutReal> default_expected{};
-  EXPECT_FALSE(griddata.get(mesh, result, "non-existent", nz));
+  EXPECT_FALSE(griddata->get(&mesh_from_options, result, "non-existent", nz));
   EXPECT_EQ(result, default_expected);
 }
+
+TEST_F(GridFromOptionsTest, CoordinatesCentre) {
+  auto coords = mesh_from_options.getCoordinates();
+
+  mesh_from_options.communicate(expected_2d);
+
+  EXPECT_TRUE(IsFieldEqual(coords->g11, expected_2d + 5.));
+  EXPECT_TRUE(IsFieldEqual(coords->g22, expected_2d + 4.));
+  EXPECT_TRUE(IsFieldEqual(coords->g33, expected_2d + 3.));
+  EXPECT_TRUE(IsFieldEqual(coords->g12, expected_2d + 2.));
+  EXPECT_TRUE(IsFieldEqual(coords->g13, expected_2d + 1.));
+  EXPECT_TRUE(IsFieldEqual(coords->g23, expected_2d));
+}
+
+TEST_F(GridFromOptionsTest, CoordinatesZlow) {
+  auto coords = mesh_from_options.getCoordinates(CELL_ZLOW);
+
+  mesh_from_options.communicate(expected_2d);
+
+  EXPECT_TRUE(IsFieldEqual(coords->g11, expected_2d + 5.));
+  EXPECT_TRUE(IsFieldEqual(coords->g22, expected_2d + 4.));
+  EXPECT_TRUE(IsFieldEqual(coords->g33, expected_2d + 3.));
+  EXPECT_TRUE(IsFieldEqual(coords->g12, expected_2d + 2.));
+  EXPECT_TRUE(IsFieldEqual(coords->g13, expected_2d + 1.));
+  EXPECT_TRUE(IsFieldEqual(coords->g23, expected_2d));
+}
+
+TEST_F(GridFromOptionsTest, CoordinatesXlowInterp) {
+  // *_xlow fields not present in options, Coordinates will be interpolated
+  // from CELL_CENTRE
+
+  // make the mesh have boundaries to avoid NaNs in guard cells after interpolating
+  mesh_from_options.createBoundaries();
+
+  auto coords = mesh_from_options.getCoordinates(CELL_XLOW);
+
+  Field2D expected_xlow = {makeField<Field2D>(
+      [](Field2D::ind_type& index) {
+        return index.x() - 0.5 + (TWOPI * index.y()) + (TWOPI * index.z() / nz) + 3;
+      },
+      &mesh_from_options)};
+
+  mesh_from_options.communicate(expected_xlow);
+
+  EXPECT_TRUE(IsFieldEqual(coords->g11, expected_xlow + 5., "RGN_NOBNDRY", this_tolerance));
+  EXPECT_TRUE(IsFieldEqual(coords->g22, expected_xlow + 4., "RGN_NOBNDRY", this_tolerance));
+  EXPECT_TRUE(IsFieldEqual(coords->g33, expected_xlow + 3., "RGN_NOBNDRY", this_tolerance));
+  EXPECT_TRUE(IsFieldEqual(coords->g12, expected_xlow + 2., "RGN_NOBNDRY", this_tolerance));
+  EXPECT_TRUE(IsFieldEqual(coords->g13, expected_xlow + 1., "RGN_NOBNDRY", this_tolerance));
+  EXPECT_TRUE(IsFieldEqual(coords->g23, expected_xlow, "RGN_NOBNDRY", this_tolerance));
+}
+
+TEST_F(GridFromOptionsTest, CoordinatesXlowRead) {
+  // *_xlow fields added to options, will be read to initialise Coordinates
+
+  // Note '(9 - x)' here because FakeMesh::GlobalX(int jx) returns jx, not a
+  // global position between 0 and 1 (in grid cells, <0 or >1 in boundaries),
+  // like a Mesh is supposed to.
```

Comment:
Ah sorry, I just did the simplest thing for `FakeMesh::GlobalX` that allowed me to write tests!

Reply:
I think it's sensible behaviour for the tests :+1: Just took me a while to figure out what was going on!